pkg/controllers/management/clusterdeploy/clusterdeploy.go

```diff
@@ -93,10 +92,6 @@ func (cd *clusterDeploy) sync(key string, cluster *v3.Cluster) (runtime.Object,
 		return nil, nil
 	}
 
-	if cluster.Status.Driver == v32.ClusterDriverImported && clusterconnected.Connected.IsFalse(cluster) {
-		return cluster, nil
-	}
-
```

Comment:
Should this check be moved to line 119 below so that the error message doesn't get set in this case? That seems to be the original intention of putting this check here (as the commit message says: https://github.com/rancher/rancher/commit/f82c084cef10ec73264473049055333ec007bbf1)

Reply:
Thanks @thedadams for calling this out. It's a good point, the original intention was indeed to avoid showing any other errors when the cluster agent is disconnected. But if the agent deploy fails for some reason, this would mean we would return nil and not retry - in that case, we'd see this issue again where imported clusters won't get registered at all. So I experimented with adding AgentDeployed condition check to the removed block for returning: 

`if cluster.Status.Driver == v32.ClusterDriverImported && v32.ClusterConditionAgentDeployedIsTrue(cluster) && clusterconnected.Connected.IsFalse(cluster) {
		return cluster, nil
	}
`

This still blocks out agent redeploys on rancher server upgrades, so I think it'd be safest to leave this as is
- to not block any agent deploys, if agent deploy itself fails, we should still set error and retry 
- once agent deploy succeeds, if for any other reason cluster disconnects for any other reason, we now have a connected condition set to show the msg

Reply:
Thanks for the thorough investigation. What you are saying makes sense to me: if there is an error redeploying the agent, then Rancher should retry.